File: src/webql/engine/transient_ffi.erl
-module(transient_ffi).

-export([run/1, start_plan/1, finish_plan/2, start_batch/1, finish_batch/3, start_step/1,
         finish_step/2]).

run(Next) ->
  case Next() of
    {ok, Task} ->
      await(Task);
    Error ->
      Error
  end.

start_plan(Next) ->
  async(fun() ->
           case Next() of
             {ok, {Initial, Batches}} -> run_batches({ok, Initial}, Batches);
             Error -> Error
           end
        end).

finish_plan(Task, Next) ->
  async(fun() ->
           case await(Task) of
             {ok, Memory} -> Next(Memory);
             Error -> Error
           end
        end).

start_batch(Next) ->
  Next().

finish_batch(Initial, Task, Merge) ->
  case Task of
    {ok, Steps} ->
      run_steps(Initial, Steps, Merge);
    Error ->
      Error
  end.

start_step(Next) ->
  async(fun() ->
           case Next() of
             {ok, Task} -> await(Task);
             Error -> Error
           end
        end).

finish_step(Task, Next) ->
  async(fun() -> Next(await(Task)) end).

run_batches({error, _} = Error, _Batches) ->
  Error;
run_batches({ok, Memory}, []) ->
  {ok, Memory};
run_batches({ok, Memory}, [Batch | Rest]) ->
  run_batches(await(Batch(Memory)), Rest).

run_steps(Memory, [], _Merge) ->
  {ok, Memory};
run_steps(Memory, [Step | Rest], Merge) ->
  case await(Step) of
    {ok, StepMemory} ->
      run_steps(Merge(Memory, StepMemory), Rest, Merge);
    Error ->
      Error
  end.

async(Fun) ->
  Pid = spawn(fun() -> wait_for_await(Fun()) end),
  {async, Pid}.

await({async, Pid}) ->
  Ref = make_ref(),
  Monitor = monitor(process, Pid),
  Pid ! {await, self(), Ref},
  receive
    {Ref, Result} ->
      demonitor(Monitor, [flush]),
      Result;
    {'DOWN', Monitor, process, Pid, Reason} ->
      exit(Reason)
  end;
await(Task) ->
  Task.

wait_for_await(Result) ->
  receive
    {await, Pid, Ref} ->
      Pid ! {Ref, Result}
  end.

File: src/webql/engine/transient_ffi.ex
defmodule TransientFfi do
  def run(next) do
    case next.() do
      {:ok, task} -> await(task)
      {:error, message} -> {:error, message}
    end
  end

  def start_plan(next) do
    async(fn ->
      case next.() do
        {:ok, {initial, batches}} ->
          Enum.reduce_while(batches, {:ok, initial}, fn batch, {:ok, memory} ->
            case await(batch.(memory)) do
              {:ok, _} = result -> {:cont, result}
              error -> {:halt, error}
            end
          end)

        error ->
          error
      end
    end)
  end

  def finish_plan(task, next) do
    async(fn ->
      case await(task) do
        {:ok, memory} -> next.(memory)
        {:error, message} -> {:error, message}
      end
    end)
  end

  def start_batch(next) do
    next.()
  end

  def finish_batch(initial, task, merge) do
    async(fn ->
      case await(task) do
        {:ok, steps} ->
          steps
          |> Enum.map(&await/1)
          |> Enum.reduce_while({:ok, initial}, fn step, {:ok, memory} ->
            case step do
              {:ok, step_memory} -> {:cont, {:ok, merge.(memory, step_memory)}}
              error -> {:halt, error}
            end
          end)

        {:error, message} ->
          {:error, message}
      end
    end)
  end

  def start_step(next) do
    async(fn ->
      case next.() do
        {:ok, task} -> await(task)
        {:error, message} -> {:error, message}
      end
    end)
  end

  def finish_step(task, next) do
    async(fn ->
      next.(await(task))
    end)
  end

  # PRIVATE FUNCTIONS
  # =================
  defp async(fun) do
    {:async,
     spawn(fn ->
       result =
         try do
           {:ok, fun.()}
         catch
           kind, reason -> {:error, kind, reason, __STACKTRACE__}
         end

       pending(result)
     end)}
  end

  defp await({:async, pid}) do
    ref = make_ref()
    monitor = Process.monitor(pid)

    send(pid, {self(), ref})

    receive do
      {^ref, {:ok, result}} ->
        Process.demonitor(monitor, [:flush])
        result

      {^ref, {:error, kind, reason, stacktrace}} ->
        Process.demonitor(monitor, [:flush])
        :erlang.raise(kind, reason, stacktrace)

      {:DOWN, ^monitor, :process, ^pid, reason} ->
        exit(reason)
    end
  end

  defp await(task), do: task

  defp pending(result) do
    receive do
      {caller, ref} ->
        send(caller, {ref, result})
        pending(result)
    end
  end
end
